### PR TITLE
EVP fetching: make operation_id part of the method identity

### PR DESCRIPTION
**Pull request by levitte**
Because the operation identity wasn't integrated with the created
methods, the following code would give unexpected results:

    EVP_MD *md = EVP_MD_fetch(NULL, "MD5", NULL);
    EVP_CIPHER *cipher = EVP_CIPHER_fetch(NULL, "MD5", NULL);

    if (md != NULL)
        printf("MD5 is a digest\n");
    if (cipher != NULL)
        printf("MD5 is a cipher\n");

The message is that MD5 is both a digest and a cipher.

Partially fixes #9106

### CODE REVIEW

**Comment by levitte** [comment]
Documentation fixed

File: crypto/evp/evp_fetch.c [hunk 1]
```diff
@@ -66,24 +65,45 @@ static OSSL_METHOD_STORE *get_default_method_store(OPENSSL_CTX *libctx)
...
         && (store = get_default_method_store(libctx)) == NULL)
         return NULL;
 
     if ((namemap = ossl_namemap_stored(libctx)) == NULL
-        || (id = ossl_namemap_add(namemap, name)) == 0)
+        || (nameid = ossl_namemap_add(namemap, name)) == 0)
         return NULL;
 
-    (void)ossl_method_store_fetch(store, id, propquery, &method);
+    methid = method_id(operation_id, nameid);
```
Comment (mattcaswell) [inline crypto/evp/evp_fetch.c]: Don't we need to check `methid` to see if it is 0 and treat it as an error if so?

File: doc/internal/man3/ossl_method_construct.pod [hunk 2]
```diff
@@ -41,13 +43,25 @@ on provider dispatch tables need to do so in exactly the same way.
 ossl_method_construct() does this while leaving it to the sub-systems
 to define more precisely how the methods are created, stored, etc.
 
+It's important to keep in mind that a method is identified by three things:
```
Comment (mattcaswell) [inline doc/internal/man3/ossl_method_construct.pod]: Typo: to keep in mind *is* that

Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: Uhmmmm, considering the sentence starts with "It's", I doubt that.

File: crypto/evp/evp_fetch.c (hunk 1 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: No need. nameid can't be zero here

File: doc/internal/man3/ossl_method_construct.pod (hunk 2 above)
Comment (mattcaswell) [reply doc/internal/man3/ossl_method_construct.pod]: Duh.

File: crypto/evp/evp_fetch.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/evp/evp_fetch.c]: But if name_id > 2^24 or operation_id > 2^8 then it will fail.

Comment (levitte) [reply crypto/evp/evp_fetch.c]: Ah

Comment (levitte) [reply crypto/evp/evp_fetch.c]: Fixed

File: crypto/core_fetch.c [hunk 3]
```diff
@@ -83,7 +83,8 @@ void *ossl_method_construct(OPENSSL_CTX *libctx, int operation_id,
     void *method = NULL;
 
     if ((method =
-         mcm->get(libctx, NULL, name, propquery, mcm_data)) == NULL) {
+         mcm->get(libctx, NULL, operation_id, name, propquery, mcm_data))
+        == NULL) {
```
Comment (mattcaswell) [inline crypto/core_fetch.c]: This indentation doesn't look quite right.

File: crypto/evp/evp_fetch.c [hunk 4]
```diff
@@ -139,14 +162,17 @@ void *evp_generic_fetch(OPENSSL_CTX *libctx, int operation_id,
...
     void *method = NULL;
 
     if (store == NULL || namemap == NULL)
         return NULL;
 
-    if ((id = ossl_namemap_number(namemap, name)) == 0
-        || !ossl_method_store_cache_get(store, id, properties, &method)) {
+    if ((nameid = ossl_namemap_number(namemap, name)) == 0
+        || !ossl_method_store_cache_get(store,
+                                        method_id(operation_id, nameid),
```
Comment (mattcaswell) [inline crypto/evp/evp_fetch.c]: method_id could fail

File: crypto/evp/evp_fetch.c [hunk 5]
```diff
@@ -167,7 +193,9 @@ void *evp_generic_fetch(OPENSSL_CTX *libctx, int operation_id,
         method = ossl_method_construct(libctx, operation_id, name,
                                        properties, 0 /* !force_cache */,
                                        &mcm, &mcmdata);
-        ossl_method_store_cache_set(store, id, properties, method);
+        nameid = ossl_namemap_number(namemap, name);
+        methid = method_id(operation_id, nameid);
```
Comment (mattcaswell) [inline crypto/evp/evp_fetch.c]: This could fail

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/core_fetch.c (hunk 3 above)
Comment (levitte) [reply crypto/core_fetch.c]: Look again

File: crypto/evp/evp_fetch.c (hunk 5 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: Right, I should check that `method != NULL`.  'cause if there's a method, we know that `method_id` did work

File: crypto/evp/evp_fetch.c (hunk 4 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: Doesn't matter much, 'cause we know that there will never be a method with calculated id being zero in the cache.  So if `method_id` fails, so will `ossl_method_store_cache_get`

File: crypto/core_fetch.c (hunk 3 above)
Comment (mattcaswell) [reply crypto/core_fetch.c]: Ah, ok

File: crypto/evp/evp_fetch.c (hunk 4 above)
Comment (mattcaswell) [reply crypto/evp/evp_fetch.c]: ok

Comment (levitte) [reply crypto/evp/evp_fetch.c]: I added a safety check anyway...

**Comment by levitte** [comment]
Not sure the appveyor errors are relevant to this PR...

**Comment by levitte** [comment]
CIs are happy now...

File: doc/internal/man3/ossl_method_construct.pod [hunk 6]
```diff
@@ -41,13 +43,25 @@ on provider dispatch tables need to do so in exactly the same way.
...
+
+=back
+
 =head2 Functions
 
 ossl_method_construct() creates a method by asking all available
-providers for a dispatch table given an C<operation_id>, an algorithm
-C<name> and a set of C<properties>, and then calling appropriate
+providers for a dispatch table given an I<operation_id>, an algorithm
+I<name> and a set of I<properties>, and then calling appropriate
```
Comment (slontis) [inline doc/internal/man3/ossl_method_construct.pod]: NIT: calling the

**Review by slontis — APPROVED** [review]
this was a small oversight :)

**Review by slontis — CHANGES_REQUESTED** [review]
Maybe add a simple test similiar to the one shown in the commit body.

**Comment by levitte** [comment]
> Maybe add a simple test similiar to the one shown in the commit body.

See #9106

**Review by slontis — APPROVED** [review]

File: doc/internal/man3/ossl_method_construct.pod (hunk 6 above)
Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: Will fix

**Comment by levitte** [comment]
Merged.

b4d3f203da6210f148b2a0c7bf5a802b55ba0e65 doc/internal/man3/ossl_method_construct.pod: follow common conventions
2ccb1b4ecab2c3ac1dc2ff81a48869a79afa7839 EVP fetching: make operation_id part of the method identity

**Pull request closed without merging**